File: L1H2/src/L1H2.cpp
#include <iostream>
#include <vector>

#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/core/core.hpp>

int main(int argc, char const *argv[])
{
    cv::Mat src;
    src = cv::imread("folder/1.png");

    if (src.empty())
    {
        std::cout << "img is empty" << std::endl;
        return -1;
    }

    cv::Mat B, G, R;
    std::vector<cv::Mat> bgrChannels;
    cv::split(src, bgrChannels);
    R = bgrChannels.at(2);
    G = bgrChannels.at(1);
    B = bgrChannels.at(0);

    cv::Mat subRB, binary;
    subRB = R - B;
    cv::threshold(subRB, binary, 100, 255, cv::THRESH_BINARY);
    cv::Mat imgBG = src.clone();
    imgBG.setTo(255, binary);

    cv::Mat grayImg;
    cv::cvtColor(imgBG, grayImg, cv::COLOR_BGR2GRAY);

    cv::Mat dilateImg, kernel;
    kernel = cv::getStructuringElement(CV_SHAPE_RECT, cv::Size(3, 3));
    cv::dilate(grayImg, dilateImg, kernel, cv::Point(-1, -1), 1);

    // find edge
    cv::Mat canny;
    cv::Canny(dilateImg, canny, 100, 200);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(canny, contours, CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE);

    // convect grey to bgr for drawing color in the image.
    cv::Mat drawing;
    cv::cvtColor(canny, drawing, cv::COLOR_GRAY2BGR);

    // contours.size() -1  is means that abandon the biggest one(the last one)
    for (size_t idx = 0; idx < contours.size() - 1; idx++)
    {
        cv::drawContours(drawing, contours, idx, cv::Scalar(0, 0, 255));
    }
    std::cout << "couter : " << contours.size() - 1 << std::endl;

    while (true)
    {
        cv::imshow("canny", canny);
        cv::imshow("drawing", drawing);
        if (cv::waitKey(1) == 27)
            break;
    }

    cv::destroyAllWindows();
    return 0;
}

File: L1H2/src/L1H2.py
import numpy as np
import cv2


def main():
    src = cv2.imread("img/1.png")

    # proc find object
    gray = cv2.cvtColor(src, cv2.COLOR_BGR2GRAY)
    canny = cv2.Canny(gray, 50, 100)

    kernel = np.ones((3, 3), np.uint8)
    dilate = cv2.dilate(canny, kernel, iterations=1)

    h, w = src.shape[:2]
    mask = np.zeros([h+2, w+2], np.uint8)
    cv2.floodFill(dilate, mask, (0, 0), (255, 255, 255))

    dilate = 255 - dilate
    contours, _ = cv2.findContours(
        dilate, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)
    print("objects : " + str(len(contours)))

    # proc blue ball
    blueBinary = cv2.inRange(src, (127, 0, 0), (255, 77, 77))
    contours, _ = cv2.findContours(
        blueBinary, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    print("blue ball : " + str(len(contours)))

    a = [cv2.minEnclosingCircle(contours[idx])[0]
         for idx in range(len(contours))]
    centers = np.uint16(a)

    # sorted_centers, centers_len = sorted(centers, key=lambda x: x[0]), len(centers)
    left = min(centers, key=lambda center: center[0])
    right = max(centers, key=lambda center: center[0])

    cv2.circle(src, tuple(left), 3, (0, 255, 0), 3)
    cv2.circle(src, tuple(right), 3, (0, 255, 0), 3)
    cv2.circle(src, (int((left[0] + right[0]) / 2),
                     int((left[1] + right[1]) / 2)), 3, (0, 0, 255), 3)

    ycrcb = cv2.cvtColor(src, cv2.COLOR_BGR2YCrCb)

    cv2.imshow("src", src)
    cv2.imshow("ycrcb", ycrcb)

    while True:

        if cv2.waitKey(1) == 27:
            break

    cv2.destroyAllWindows()


if __name__ == '__main__':
    main()
